### ISSUE TITLE
gRPC generates "data corruption" errors while using bigtable::MutationBatcher

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#4018]
**Does this issue affect the google-cloud-cpp project?**

Maybe, it could be lower in the stack (gRPC? OpenSSL?), but I am creating this bug to track it down and maybe repro the problem.

**What component of `google-cloud-cpp` is this related to?**

Bigtable.

**Describe the bug**

When using `bigtable::MutationBatcher` gRPC reports:

```
E0506 18:42:58.378414970  198237 ssl_transport_security.cc:483] Corruption detected.
E0506 18:42:58.394900909  198237 ssl_transport_security.cc:459] error:14187180:SSL routines:ssl_do_config:bad value
E0506 18:42:58.395328567  198237 ssl_transport_security.cc:459] error:14187180:SSL routines:ssl_do_config:bad value
E0506 18:42:58.395932424  198237 secure_endpoint.cc:208]     Decryption error: TSI_DATA_CORRUPTED
E0506 18:42:58.636754212  198227 ssl_transport_security.cc:483] Corruption detected.
```

**To Reproduce**

I have a repro in a [branch](https://github.com/coryan/google-cloud-cpp/tree/repro-mutation-batcher-data-corruption)

**Expected behavior**

No errors (or normal gRPC errors!)

**Operating system:**

Linux. Something based on Debian (9 I believe).

**What compiler and version are you using?**

g++ --version
g++ (Debian 9.2.1-28) 9.2.1 20200203
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**What version of `google-cloud-cpp` are you using?**

https://github.com/coryan/google-cloud-cpp/tree/repro-mutation-batcher-data-corruption

**Additional context**

It seems that there is only an issue when running multiple threads in the completion queue event loop.

**Comment by coryan** [comment]
It is not clear if these messages cause actual problems, my repro (linked above) has zero application-level errors even with many of these messages in the log.

**Comment by coryan** [comment]
Using non-idempotent operations makes this issue visible at the application level.

**Comment by coryan** [comment]
I am not sure what this means, but building with Bazel makes the problems go away.  I am beginning to suspect OpenSSL.

**Comment by coryan** [comment]
I tested with BoringSSL and CMake and the problems go away too.  I think we have established (to the preponderance of evidence) that OpenSSL has a bug (and/or gRPC does not use OpenSSL correctly in multi-threaded environments, but that is less likely).

The problem has become: how do we support BoringSSL?  It is easy enough to describe how to use it:

- Download and compile BoringSSL manually, install it somewhere by manually copying headers and libraries.
- Compile everything with that version of BoringSSL: gRPC, the `google-cloud-cpp` libraries, etc.  Most likely by using the `OPENSSL_ROOT_DIR` environment variable (or CMake variable) to point CMake to the version you just compiled.

But how do we test it?
- What additions do we want in the build matrix?
- Do we want to test "manually build and install in /usr/local" using Dockerfiles?
- Do we want to test "use a super build"?
- Do we want to test macOS, Windows, and multiple Linux distros?

**Comment by coryan** [comment]
I am not actively working on this problem.  I think we need to give BoringSSL better support, but I am not sure how.

**Comment by coryan** [comment]
Seems related (I would say caused by): grpc/grpc#23144

**Comment by devbww** [comment]
> Seems related (I would say caused by): grpc/grpc#23144

Given that we're waiting on an external bug, lets move this issue to P3.

**Comment by coryan** [comment]
The `external` label can be used to tell the SLO tool that something is blocked on an external bug.

**Comment by devbww** [comment]
Excellent.  Thanks.

**Comment by coryan** [comment]
Seems like the underlying bugs in gRPC are fixed.  I guess we can close this once we pick up a version of gRPC that contains the fixes mentioned in grpc/grpc#23144

**Comment by coryan** [comment]
Closing, the underlying bug in gRPC is fixed on the latest release *and* we are using that in our tests.
